src/controllers/scripting/legacy/controllerscriptenginelegacy.cpp

```diff
@@ -139,15 +343,87 @@ bool ControllerScriptEngineLegacy::initialize() {
+                    QDirIterator::Subdirectories);
+            while (it.hasNext()) {
+                watchFilePath(it.next());
+            }
+            watchFilePath(path);
+            auto pQmlEngine = std::dynamic_pointer_cast<QQmlEngine>(m_pJSEngine);
+            pQmlEngine->addImportPath(path);
+            qCWarning(m_logger) << pQmlEngine->importPathList();
+        }
+    } else if (!m_modules.isEmpty()) {
+        qCWarning(m_logger) << "Controller mapping has QML library definitions but no "
+                               "QML files to use it. Ignoring.";
+    }
+
+    // If we encounter a failure while loading a scene, we will need to properly
+    // stop the screen threads before shutting down.
+    bool sceneBindingHasFailure = false;
+#endif
     for (const LegacyControllerMapping::ScriptFileInfo& script : std::as_const(m_scriptFiles)) {
-        if (!evaluateScriptFile(script.file)) {
-            shutdown();
-            return false;
+#ifdef MIXXX_USE_QML
+        if (script.type == LegacyControllerMapping::ScriptFileInfo::Type::JAVASCRIPT) {
+#endif
+            if (!evaluateScriptFile(script.file)) {
+                shutdown();
+                return false;
+            }
+            if (!script.identifier.isEmpty()) {
+                m_scriptFunctionPrefixes.append(script.identifier);
+            }
+#ifdef MIXXX_USE_QML
+        } else {
+            if (script.identifier.isEmpty()) {
+                while (!availableScreens.isEmpty()) {
+                    QString screenIdentifier(availableScreens.firstKey());
+                    if (!bindSceneToScreen(script,
+                                screenIdentifier,
+                                availableScreens.take(screenIdentifier))) {
+                        sceneBindingHasFailure = true;
+                    }
+                }
+            } else {
+                if (!availableScreens.contains(script.identifier)) {
+                    qCCritical(m_logger) << "Not screen" << script.identifier << "found!";
+
+                    sceneBindingHasFailure = true;
+                    break;
+                }
+                if (!bindSceneToScreen(script,
+                            script.identifier,
+                            availableScreens.take(script.identifier))) {
+                    sceneBindingHasFailure = true;
+                }
+            }
         }
-        if (!script.functionPrefix.isEmpty()) {
-            m_scriptFunctionPrefixes.append(script.functionPrefix);
+    }
+
+    if (!availableScreens.isEmpty()) {
+        if (!sceneBindingHasFailure) {
+            qCWarning(m_logger)
+                    << "Found screen with no QML scene able to run on it. Ignoring"
+                    << availableScreens.size() << "screens";
+        }
+
+        while (!availableScreens.isEmpty()) {
+            auto orphanScreen = availableScreens.take(availableScreens.firstKey());
+            std::move(orphanScreen)->deleteLater();
```

Comment:
`std::move` will do nothing here.

Reply:
IIUC, it will extract the pointer from the unique_ptr, so it doesn't get deallocated when `orphanScreen` goes out of scope. Instead, it will be deallocated by the evenloop owning this object "later".

Reply:
no. `std::move` is really only a cast, no instructions get emitted at all. In this case, I'm pretty sure `orphanScreen` will get deallocated right then at there at the end of the scope.

Reply:
Hopefully, this `swap` and custom deleter approach works better.

Reply:
we'll see once you push the commit :wink:

Reply:
Done!